File: src/Sketch.java
import processing.core.PApplet;

public class Sketch extends PApplet{
    public static Object[] objects;
    public static void main(String[] args) {
        String[] processingArgs = {"Sketch"};
        Sketch sketch = new Sketch();
        PApplet.runSketch(processingArgs,sketch);
    }

    @Override
    public void settings() {
        size(800,800);
    }

    public void setup() {
            var polyPos = new Vector[] {new Vector(410,390), new Vector(390,390), new Vector(390,410), new Vector(410,410)};
        objects = new Object[] {
            new Polygon(polyPos, new Addons[] {new Gravity(),new Collision()}, color(123,234,234), false),
            // new Square(new Vector(0,height),new Vector(width,height),new Vector(width,height/4*3), new Vector(0,height/4*3),new Addons[] {},color(0,0,0), true),
            new Square(new Vector(0,height),new Vector(width,height),new Vector(width,height), new Vector(0,400),new Addons[] {},color(0,0,0), true),
            new Square(new Vector(0,height),new Vector(width,height),new Vector(width,400), new Vector(0,height),new Addons[] {},color(0,0,0), true),
            new Square(new Vector(0,0),new Vector(width,0),new Vector(width,0), new Vector(0,400),new Addons[] {},color(0,0,0), true),
            new Square(new Vector(0,0),new Vector(width,0),new Vector(width,400), new Vector(0,0),new Addons[] {},color(0,0,0), true),
            // new Ball(new Vector(10,100),10,new Addons[] {new Gravity(), new Collision()},color(100,100,100)), 
            // new Ball(new Vector(10,100),50,new Addons[] {},color(100,100,100)),
            // new Square(new Vector(10,10), new Vector(100,0), new Vector(100,100), new Vector(0,100), new Addons[] {}, color(100,100,100)),
            // new Square(new Vector(15,15), 20,20, new Addons[] {new Gravity(),new Collision()}, color(123,32,12,50))
        };
    }
    
    public void draw() {
        background(220);
        for(Object object: objects) {
            object.tick(objects);
        }

        for(Object object: objects) {
            object.draw(g);
        }

    }
}

File: sketch/sketch.ts
let simulation: Simulation;

let sketch = (p: p5) => {
  p.setup = () => {
    let width = 500;
    let height = 500;
    p.createCanvas(width, height);
    p.frameRate(60);
    simulation = new Simulation(width, height);
    for (let i = 0; i < 100; i++) {
      simulation.addBall(10);
    }
  };

  p.draw = () => {
    p.background(220);
    simulation.render();
  };
};

function setup() {
  let width = 500;
  let height = 500;
  createCanvas(width, height);
  frameRate(60);
  simulation = new Simulation(width, height);
  for (let i = 0; i < 100; i++) {
    simulation.addBall(10);
  }
}

function draw() {
  background(220);
  simulation.update();
  simulation.render();
}
class Simulation {
  width: number;
  height: number;
  balls: Ball[] = [];
  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
    this.balls = [];
  }

  addBall(radius: number) {
    let diameter = radius * 2;

    let minX = diameter;
    let minY = diameter;
    let maxX = this.width - diameter;
    let maxY = this.height - diameter;

    let x = Math.random() * (maxX - minX) + minX;
    let y = Math.random() * (maxY - minY) + minY;
    let direction = Math.random() * Math.PI * 2;
    let ballSpeed = 1;

    this.balls.push(new Ball(x, y, radius, direction, ballSpeed));
  }

  update() {
    this.balls.forEach((ball) => {
      ball.move();
    });
    simulation.checkForCollisionsIndexed();
    simulation.testForWalls();
  }

  testForWalls() {
    for (let ball of this.balls) {
      // too far right
      if (ball.x + ball.radius > this.width) {
        ball.dx *= -1;
        ball.x = this.width - ball.radius;
        continue;
      }

      // too far left
      if (ball.x - ball.radius < 0) {
        ball.dx *= -1;
        ball.x = ball.radius;
        continue;
      }

      // too far down
      if (ball.y + ball.radius > this.height) {
        ball.dy *= -1;
        ball.y = this.height - ball.radius;
        continue;
      }

      // too far up
      if (ball.y - ball.radius < 0) {
        ball.dy *= -1;
        ball.y = ball.radius;
      }
    }
  }

  checkForCollisionsIndexed() {
    // create our new index
    let index = new Flatbush(this.balls.length);

    // add them
    for (const ball of this.balls) {
      index.add(ball.bbox[0], ball.bbox[1], ball.bbox[2], ball.bbox[3]);
    }

    // prep the index
    index.finish();

    for (const first of this.balls) {
      const collisions = index.search(first.bbox[0], first.bbox[1], first.bbox[2], first.bbox[3]);

      for (const idx of collisions) {
        const second = this.balls[idx];

        if (first === second) continue;

        if (first.isCollidingWith(second)) {
          this.collide(first, second);
        }
      }
    }
  }

  checkForCollisions() {
    // this could be much more efficient, but you need ~1,000+ balls for it
    // to start to matter, so I kept it simple
    for (const first of this.balls) {
      for (const second of this.balls) {
        // skip if same ball
        if (first === second) continue;
        if (first.isCollidingWith(second)) {
          this.collide(first, second);
        }
      }
    }
  }

  collide(first: Ball, second: Ball) {
    // first we determine if they're moving toward each other
    // this allows overlaps going in opposite directions to "resolve"
    // and not stick to each other
    const dx = first.x - second.x;
    const dy = first.y - second.y;
    const vx = second.dx - first.dx;
    const vy = second.dy - first.dy;
    const dot = dx * vx + dy * vy;

    // if going toward each other, bounce 'em
    if (dot > 0) {
      const cr = first.mass + second.mass;
      const dr = first.mass - second.mass;

      // determine our new velocities
      const vx1 = (first.dx * dr + 2 * second.mass * second.dx) / cr;
      const vy1 = (first.dy * dr + 2 * second.mass * second.dy) / cr;
      const vx2 = (second.dx * -dr + 2 * first.mass * first.dx) / cr;
      const vy2 = (second.dy * -dr + 2 * first.mass * first.dy) / cr;

      // set them
      first.dx = vx1;
      first.dy = vy1;
      second.dx = vx2;
      second.dy = vy2;

      // trigger a move for both so they stop overlapping immediately
      first.move();
      second.move();
    }
  }
  render() {
    for (const ball of this.balls) {
      circle(ball.x, ball.y, ball.radius);
    }
  }
}
class Ball {
  x: number;
  y: number;
  radius: number;
  mass: number;
  direction: number;
  speed: number;
  dx: number;
  dy: number;
  constructor(x: number, y: number, radius: number, direction: number, ballSpeed: number) {
    this.x = x;
    this.y = y;
    this.radius = radius;
    this.mass = radius;
    this.direction = direction;
    this.speed = ballSpeed;
    this.dx = this.speed * Math.cos(this.direction);
    this.dy = this.speed * Math.sin(this.direction);
  }

  move() {
    this.x += this.speed;
    this.y += this.speed;
  }

  isCollidingWith(ball: Ball) {
    const dx = this.x - ball.x;
    const dy = this.y - ball.y;
    const dr = this.radius + ball.radius;
    // don't use the square root because it's 1) slower, and 2) not needed for collision detection
    const squaredDistance = Math.pow(dx, 2) + Math.pow(dy, 2);

    return squaredDistance < Math.pow(dr, 2);
  }

  get bbox() {
    return [this.x - this.radius, this.y - this.radius, this.x + this.radius, this.y + this.radius];
  }
}
